dozer-api/src/grpc/mod.rs

```diff
@@ -11,23 +11,143 @@ mod shared_impl;
 pub mod typed;
 pub mod types_helper;
 
+use std::io;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
 use bytes::Bytes;
 pub use client_server::ApiServer;
 use dozer_types::errors::internal::BoxedError;
-use dozer_types::tonic::transport::server::{Router, Routes, TcpIncoming};
-use futures_util::{
-    stream::{AbortHandle, Abortable, Aborted},
-    Future,
+use dozer_types::tonic::transport::server::{
+    Connected, Router, Routes, TcpConnectInfo, TcpIncoming,
 };
+use futures_util::Future;
+use futures_util::StreamExt;
 pub use grpc_web_middleware::enable_grpc_web;
 use http::{Request, Response};
+use hyper::server::conn::AddrStream;
 use hyper::Body;
+use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
 use tower::{Layer, Service};
 
+use crate::shutdown::ShutdownReceiver;
+
+#[derive(Debug)]
+enum ShutdownAddrStream<F> {
```

Comment:
This can be simplified a lot by making this
```rust
struct ShutdownAddrStream<F> {
    inner: AddrStream,
    shutdown: Option<F>,
}
```

Reply:
If you want to be clearer about the state, it can be
```rust
enum State {
    Alive{ shutdown: F },
    Shutdown
}

struct ShutdownAddrStream<F> {
    inner: AddrStream,
    state: State,
}
```

but i think just using an `Option` with maybe a short comment is plenty clear in this case

Reply:
While I was making this change, I realized that we didn't pool `AddrStream` shutdown until completion, making the disconnecting not so graceful. See the latest commit.